src/lang/checkers/SanityChecker.ml

```diff
@@ -0,0 +1,120 @@
+      in
+      let (e, _) = List.fold_left (fun (e, dlist) i ->
+        let count = list_count ilist i in
+        if count > 1 && not (List.exists (fun x -> get_id x = get_id i) dlist) then
+          (e ^ Core.sprintf "[%s] Identifier %s used more than once\n"
+                            (get_loc_str @@ gloc @@ get_rep i) (get_id i)), i::dlist
+        else (e, dlist)
+      ) ("", []) ilist
+      in
+        e
+    in
+
+    (* No repeating names for params. *)
+    let e = check_duplicate_ident ER.get_loc (List.map (fun (i, _) -> i) contr.cparams) in
+    (* No repeating field names. *)
+    let e = e ^ check_duplicate_ident ER.get_loc (List.map (fun (i, _, _) -> i) contr.cfields) in
+    (* No repeating transition names. *)
+    let e = e ^ check_duplicate_ident SR.get_loc (List.map (fun t -> t.tname) contr.ctrans) in
+    (* No repeating transition parameter names. *)
+    let e = List.fold_left
+      (fun e t -> e ^ check_duplicate_ident ER.get_loc (List.map (fun (i, _) -> i) t.tparams))
+       e contr.ctrans 
+    in
+
+    (* Message literals must either be for "send" or "event" and well formed. *)
+    let check_message b msg e =
+      (* Use location of "b" to represent the location of msg. *)
+      let eloc_str = Core.sprintf "[%s] " (get_loc_str @@ ER.get_loc @@ get_rep b) in
+
+      (* No repeating message field. *)
+      let e' = check_duplicate_ident SR.get_loc (List.map (fun (s, _) -> SR.mk_id_string s) msg) in
+      (* Prepend message location as the literal itself has no location info. *)
+      let e = if e' <> "" then e ^ eloc_str ^ e' else e in
+
+      (* Either "_tag" or "_eventname" must be present. *)
+      let e = if (List.exists (fun (s, _) -> s = tag_label) msg)
+      then
+        (* This is a "send" Message. Ensure "_amount" and "_recipient" provided. *)
+        if List.exists (fun (s, _) -> s = amount_label) msg &&
+           List.exists (fun (s, _) -> s = recipient_label) msg 
+        then e 
+        else e ^ eloc_str ^ "Missing " ^ amount_label ^ " or " ^ recipient_label ^ " in Message\n"
+      else
+        (* This is an "event" message, and must have "_eventname" field. *)
+        if List.exists (fun (s, _) -> s = eventname_label) msg
+        then e
+        else e ^ eloc_str ^ "Missing " ^ eventname_label ^ " field in message\n"
+      in
+        pure e (* as required by "fold_over_messages" *)
+    in
+    let%bind e = fold_over_messages contr ~init:e ~f:check_message in
+
+    (* Transition parameters cannot have names as that of implicit ones. *)
+    let e = List.fold_left (fun e t -> 
+      match List.find_opt (fun (s, _) -> get_id s = amount_label || get_id s = sender_label) t.tparams with
+      | Some (s, _) ->
+        e ^ Core.sprintf "[%s] Paramter %s in transition %s cannot be explicit.\n" 
+        (get_loc_str @@ SR.get_loc @@ get_rep t.tname) (get_id s) (get_id t.tname)
+      | None -> e
+      ) e contr.ctrans in
+
```

Comment:
As a general observation: We now have implicit fields and parameters hardcoded here and ind the typechecker, so if we decide to add an implicit field later on, we have to add it in multiple places. Is it possible to move the names of implicit fields and parameters to somewhere like `ContractUtil.ml`, so that they only exist in one place?

Reply:
I'm not sure I follow. The names _are_ in `ContractUtil.ml`. Do you mean that the check itself be moved?

Reply:
No, I mean that the fact that these are the implicit fields/parameters should be moved. Maybe have a list of fields/parameters in `ContractUtil.ml` that we can perform the check against.

In any case it's a separate issue, so it doesn't need to be fixed as part of this PR.